benchmarks/dbscan/dbscan.cpp

```diff
@@ -0,0 +1,178 @@
+/****************************************************************************
+ * Copyright (c) 2017-2022 by the ArborX authors                            *
+ * All rights reserved.                                                     *
+ *                                                                          *
+ * This file is part of the ArborX library. ArborX is                       *
+ * distributed under a BSD 3-clause license. For the licensing terms see    *
+ * the LICENSE file in the top-level directory.                             *
+ *                                                                          *
+ * SPDX-License-Identifier: BSD-3-Clause                                    *
+ ****************************************************************************/
+
+#include "dbscan.hpp"
+
+#include <ArborX_DBSCAN.hpp>
+#include <ArborX_Version.hpp>
+
+#include <Kokkos_Core.hpp>
+
+#include <boost/program_options.hpp>
+
+#include <cstdlib>
+#include <fstream>
+
+void getDataDimensions(std::string const &filename, bool binary,
+                       int &num_points, int &dim)
+{
+  std::ifstream input;
+  if (!binary)
+    input.open(filename);
+  else
+    input.open(filename, std::ifstream::binary);
+  ARBORX_ASSERT(input.good());
+
+  if (!binary)
+  {
+    input >> num_points;
+    input >> dim;
+  }
+  else
+  {
+    input.read(reinterpret_cast<char *>(&num_points), sizeof(int));
+    input.read(reinterpret_cast<char *>(&dim), sizeof(int));
+  }
+  input.close();
+}
+
+namespace ArborX::DBSCAN
+{
+// This function is required for Boost program_options to be able to use the
+// Implementation enum.
+std::istream &operator>>(std::istream &in, Implementation &implementation)
+{
+  std::string impl_string;
+  in >> impl_string;
+
+  if (impl_string == "fdbscan")
+    implementation = ArborX::DBSCAN::Implementation::FDBSCAN;
+  else if (impl_string == "fdbscan-densebox")
+    implementation = ArborX::DBSCAN::Implementation::FDBSCAN_DenseBox;
+  else
+    in.setstate(std::ios_base::failbit);
+
+  return in;
+}
+
+// This function is required for Boost program_options to use Implementation
+// enum as the default_value().
+std::ostream &operator<<(std::ostream &out,
+                         Implementation const &implementation)
+{
+  switch (implementation)
+  {
+  case ArborX::DBSCAN::Implementation::FDBSCAN:
+    out << "fdbscan";
+    break;
+  case ArborX::DBSCAN::Implementation::FDBSCAN_DenseBox:
+    out << "fdbscan-densebox";
+    break;
+  }
+  return out;
+}
+} // namespace ArborX::DBSCAN
```

Comment:
Consider moving these to their own source file

Reply:
I thought about it, but did not want to create another file.

Reply:
These were removed after the switch to `std::string`.